core/src/main/java/edu/wpi/grip/core/sources/VideoFileSource.java

```diff
@@ -0,0 +1,198 @@
+    if (isNewFrame.compareAndSet(true, false)) {
+      // New frame, update outputs
+      synchronized (workingMat) {
+        workingMat.copyTo(imageSocket.getValue().get());
+      }
+      imageSocket.setValue(imageSocket.getValue().get()); // force the socket to update
+      return true;
+    } else {
+      // No new frame, no update
+      return false;
+    }
+  }
+
+  @Override
+  public Properties getProperties() {
+    Properties p = new Properties();
+    p.put("path", path);
+    return p;
+  }
+
+  private static String getVideoFormat(String path) {
+    return path.substring(path.lastIndexOf('.') + 1, path.length());
+  }
+
+  @Override
+  public void initialize() throws IOException {
+    try {
+      frameGrabber = new FFmpegFrameGrabber(path);
+      frameGrabber.setFormat(getVideoFormat(path));
+      frameGrabber.start();
+      final double fps = frameGrabber.getFrameRate();
+      fpsSocket.setValue(fps);
+      captureThread = new VideoCaptureThread(fps, path);
+      captureThread.start();
+    } catch (FrameGrabber.Exception e) {
+      throw new IOException("Could not open video file " + path, e);
+    }
+  }
+
+  @Subscribe
+  public void onSourceRemoved(SourceRemovedEvent sourceRemovedEvent) {
+    if (sourceRemovedEvent.getSource() == this) {
+      try {
+        if (captureThread != null) {
+          captureThread.quit();
+        }
+        frameGrabber.stop();
+      } catch (FrameGrabber.Exception e) {
+        getExceptionWitness().flagException(e, "Exception when stopping frame grabber");
+      } finally {
+        eventBus.unregister(this);
+      }
+    }
+  }
+
+  public interface Factory {
+    VideoFileSource create(File file);
+
+    VideoFileSource create(Properties props);
+  }
+
+  private class VideoCaptureThread extends Thread {
+
+    private final double fps;
+    private final String path;
+    private boolean doRun = true;
+
+    private VideoCaptureThread(double fps, String path) {
+      super("VideoFileCaptureThread-" + path);
+      setDaemon(true);
+      this.fps = fps;
+      this.path = path;
+    }
+
+    @Override
+    public void run() {
+      while (doRun) {
+        try {
+          Frame frame = frameGrabber.grabFrame();
+          if (frame == null) {
+            // End of the video file, loop back to the first frame
+            frameGrabber.setFrameNumber(0);
+            continue;
+          }
+          Mat m = converter.convert(frame);
+          if (m == null) {
+            // Try again (I have no idea why this happens)
+            continue;
+          }
+          synchronized (workingMat) {
+            m.copyTo(workingMat);
+          }
+          m.release();
+          isNewFrame.set(true);
+          eventBus.post(new SourceHasPendingUpdateEvent(VideoFileSource.this));
+          Thread.sleep((long) (1e3 / fps));
+        } catch (InterruptedException | FrameGrabber.Exception e) {
+          getExceptionWitness().flagException(e);
```

Comment:
Handle the `InterruptedException` like an `InterruptedException` should be handled.
http://www.yegor256.com/2015/10/20/interrupted-exception.html